feat: Audit secrets outside an environment

aldur opened the pull request.

woodruffw commented:
Hey @aldur, thanks for taking a stab at this! I'll do a high-level review later today or tomorrow.

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,80 @@
+        Ok(Self)
+    }
+
+    fn audit_raw<'w>(
```
Unless I'm missing something, I don't think you want `audit_raw` here -- that's only for low-level audits like `unredacted-secrets`.

Since you're operating purely on the YAML here and are focused on step-level findings, I think you want `audit_step` here.

woodruffw replied:
Or rather, `audit_normal_job`, since you're testing the job's `environment`.

aldur replied:
Thank you! I went for the `audit_step` and access the `environment` by going up the hierarchy. Addressed in `1d5b2df`.

woodruffw commented on tests/integration/test-data/secrets-outside-environment.yml:
Could you add a snapshot test for this? See `snapshot.rs` for how it's done for other audits :slightly_smiling_face:

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,80 @@
+                            } => match env {
+                                LoE::Expr(_) => {
+                                    // TODO: Implement this.
+                                    panic!("We don't handle Expr yet!")
```
Just noting this TODO :slightly_smiling_face: -- for reference, the `template-injection` or `unredacted-secrets` audits might be a good source for using the expression parsing APIs.

aldur replied:
Thank you! Addressed in `1d5b2df`.

aldur commented:
Thanks for the comments! I have addressed them, added a snapshot and had to modify `use-trusted-publishing.yml` in `6b963b6` for `cargo insta test` to pass. 

I have also some `.snap.new` files in my cwd, should I add them to this PR?

woodruffw commented:
> I have also some `.snap.new` files in my cwd, should I add them to this PR?

You should run `cargo insta test --review` and then accept them, and then check _those_ files in 🙂

aldur commented:
> > I have also some `.snap.new` files in my cwd, should I add them to this PR?
> 
> You should run `cargo insta test --review` and then accept them, and then check _those_ files in 🙂

Done, thank you!

woodruffw commented:
Thanks @aldur, this is looking really good! 

One thing I'm realizing, based on the changes to the other tests: this might be slightly too conservative as a default audit, at least for now, since a lot of people have relatively simple CI/CD setups with 1-2 workflows where an environment wouldn't provide much isolation.

Given that, I'm thinking that it might make sense to mark this as a "pedantic" audit for now, i.e. by calling `.persona(Persona::Pedantic)` on each finding produced. How would you feel about that? IMO that would allow people to opt into the behavior here while not being too disruptive, and we could make it a "regular" audit in the future (or make it enableable in the regular persona once #617 is fleshed out more).

aldur commented:
> Thanks @aldur, this is looking really good!
> 
> One thing I'm realizing, based on the changes to the other tests: this might be slightly too conservative as a default audit, at least for now, since a lot of people have relatively simple CI/CD setups with 1-2 workflows where an environment wouldn't provide much isolation.
> 
> Given that, I'm thinking that it might make sense to mark this as a "pedantic" audit for now, i.e. by calling `.persona(Persona::Pedantic)` on each finding produced. How would you feel about that? IMO that would allow people to opt into the behavior here while not being too disruptive, and we could make it a "regular" audit in the future (or make it enableable in the regular persona once #617 is fleshed out more).

Done! Sorry it took me some time to get back to this :)

woodruffw commented:
> Done! Sorry it took me some time to get back to this :)

No problem at all! I'll do another review tonight.

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,83 @@
+
+audit_meta!(
+    SecretsOutsideEnvironment,
+    "secrets-outside-environment",
```
I'm trying to make new audit names a bit shorter, so some ideas here:

1. `secret-outside-env`
2. `secret-without-env`
3. `secret-missing-env`
4. `secret-no-env` (shortest, but maybe not easy to understand)

Curious if you have any other naming thoughts as well :slightly_smiling_face:

aldur replied:
Picked number 2 and fixed in `96c90ac`.

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,83 @@
+        step: &crate::models::Step<'w>,
+        findings: &mut Vec<crate::finding::Finding<'w>>,
+    ) -> anyhow::Result<()> {
+        if s.contains("secrets") {
```
I *think* we need to exclude `secrets.github_token` here, since that's a "latent" secret that's present regardless of any environment.

aldur replied:
Yep that makes sense! Fixed it [here](b5dab62).

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,83 @@
+                working_directory: _,
+            } => match env {
+                LoE::Expr(e) => {
+                    Self::check_secrets_access(e.as_bare(), step, &mut findings)?;
```
I think this check needs to be a little more discerning: `e` needs to be parsed into an `Expr`, which then needs to be walked recursively to visit each `Expr::Context` and match _that_ on the `secrets` namespace. 

Without that, we'll end up with (admittedly goofy) false positives like:

```yaml
with:
  yolo: ${{ 'secrets' }}
```

Writing the AST walker is currently a bit annoying/manual; if you want to wait until I add a `Visitor` API that might make your life a bit easier (I plan on doing that pretty soon) :slightly_smiling_face:

aldur replied:
I like that! It will also increase the audit confidence. I am ok with waiting for the `Visitor` API, just give me a ping once ready and I'll put it to work :)

woodruffw commented on src/audit/secrets_outside_environment.rs:
```diff
@@ -0,0 +1,83 @@
+                eenv = with;
+            }
+            StepBody::Run {
+                run: _,
```
Note: I'm okay with skipping `run:` blocks for an MVP here, but we _will_ need to extract expressions from those and scan for secrets as well at some point.

aldur replied:
Makes sense! I naively thought that wouldn't be required, would this be an example?

```
- name: Clean install dependencies and build
  run: |
    npm ci
    PASSWORD=${{ secrets.password}} npm run build
```

aldur commented:
Hi there! Any news on the `Visitor` API? I can give this another shot and get it merged.
